File: SmartSight/SmartSight/SmartSightPlusPlus/markerDetector.cpp
#include "markerDetector.hpp"
#include <opencv2\core\types_c.h>
#include <opencv2\imgproc\types_c.h>

markerDetector::markerDetector(CameraCalibration calibration) :m_minContourLengthAllowed(100), markerSize(100, 100), cellSize(100 / 7)
{
	cv::Mat(3, 3, CV_32F, const_cast<float*>(&calibration.getIntrinsic().data[0])).copyTo(camMatrix);
	cv::Mat(4, 1, CV_32F, const_cast<float*>(&calibration.getDistorsion().data[0])).copyTo(distCoeff);

	m_markerCorners3d.push_back(cv::Point3f(-0.5f, -0.5f, 0));
	m_markerCorners3d.push_back(cv::Point3f(+0.5f, -0.5f, 0));
	m_markerCorners3d.push_back(cv::Point3f(+0.5f, +0.5f, 0));
	m_markerCorners3d.push_back(cv::Point3f(-0.5f, +0.5f, 0));

	m_markerCorners2d.push_back(cv::Point2f(0, 0));
	m_markerCorners2d.push_back(cv::Point2f(static_cast<float>(markerSize.width - 1), 0));
	m_markerCorners2d.push_back(cv::Point2f(static_cast<float>(markerSize.width - 1), static_cast<float>(markerSize.height - 1)));
	m_markerCorners2d.push_back(cv::Point2f(0, static_cast<float>(markerSize.height - 1)));

	templateMarker = new int[25]{ 1, 1, 1, 0, 1,
								  0, 0, 0, 1, 0,
								  1, 1, 1, 1, 0,
								  1, 1, 1, 1, 0,
								  1, 1, 1, 0, 0 };
}

void markerDetector::findMarkers(const cv::Mat& input, bool showImg)
{
	prepareImage(input, grayscalse);

	performThreshold(grayscalse, thresholdImg);

	findContours(thresholdImg, contours, grayscalse.cols / 5);

	findCandidates(contours, detectedMarkers);

	goodMarkers = recognizeMarkers(grayscalse, detectedMarkers, showImg);

	estimatePosition(goodMarkers);
}

void markerDetector::prepareImage(const cv::Mat& bgrMat, cv::Mat& grayscale) const
{
	cv::cvtColor(bgrMat, grayscale, CV_BGR2GRAY);
}

void markerDetector::performThreshold(const cv::Mat& grayscale, cv::Mat& thresholdImg) const
{
	cv::adaptiveThreshold(grayscale, thresholdImg, 255, cv::ADAPTIVE_THRESH_GAUSSIAN_C, cv::THRESH_BINARY_INV, 7, 7);
}

void markerDetector::findContours(const cv::Mat& thresholdImg, std::vector<std::vector<cv::Point>>& contours, int minContourPointsAllowed)
{
	std::vector<std::vector<cv::Point>> allContours;
	cv::findContours(thresholdImg, allContours, CV_RETR_LIST, CV_CHAIN_APPROX_NONE);

	contours.clear();

	for (size_t i = 0; i < allContours.size(); i++)
	{
		size_t contourSize = allContours[i].size();
		if (contourSize > minContourPointsAllowed)
		{
			contours.push_back(allContours[i]);
		}
	}
}

float perimeter(const std::vector<cv::Point2f>& a)
{
	float sum = 0, dx, dy;

	for (size_t i = 0; i < a.size(); i++)
	{
		size_t i2 = (i + 1) % a.size();

		dx = a[i].x - a[i2].x;
		dy = a[i].y - a[i2].y;

		sum += sqrt(dx * dx + dy * dy);
	}

	return sum;
}


void markerDetector::findCandidates(const std::vector<std::vector<cv::Point>>& contours, std::vector<std::vector<cv::Point2f>>& marks)
{
	std::vector<cv::Point> approxCurve;
	std::vector<std::vector<cv::Point2f>> possibleMarkers;

	// For each contour, analyze if it is a parallelepiped likely to be the	marker
	for (size_t i = 0; i < contours.size(); i++)
	{
		//Approximate to a polygon
		double eps = contours[i].size() * 0.05;
		cv::approxPolyDP(contours[i], approxCurve, eps, true);

		// We interested only in polygons that contains only four points
		if (approxCurve.size() != 4)
			continue;

		// And they have to be convex
		if (!cv::isContourConvex(approxCurve))
			continue;

		// Ensure that the distance between consecutive points is large enough
		float minDist = std::numeric_limits<float>::max();
		for (int i = 0; i < 4; i++)
		{
			cv::Point side = approxCurve[i] - approxCurve[(i + 1) % 4];
			float squaredSideLength = static_cast<float>(side.dot(side));
			minDist = std::min(minDist, squaredSideLength);
		}

		// Check that distance is not very small
		if (minDist < m_minContourLengthAllowed)
			continue;

		// All tests are passed. Save marker candidate:
		std::vector<cv::Point2f> m;
		for (int i = 0; i < 4; i++)
			m.push_back(cv::Point2f(static_cast<float>(approxCurve[i].x), static_cast<float>(approxCurve[i].y)));

		// Sort the points in anti-clockwise order
		// Trace a line between the first and second point.
		// If the third point is at the right side, then the points are anticlockwise
		cv::Point v1 = m[1] - m[0];
		cv::Point v2 = m[2] - m[0];

		double o = (v1.x * v2.y) - (v1.y * v2.x);
		if (o < 0.0) //if the third point is in the left side,	then sort in anti - clockwise order
			std::swap(m[1], m[3]);
		possibleMarkers.push_back(m);
	}

	// Remove these elements which corners are too close to each other.
	// First detect candidates for removal:
	std::vector< std::pair<int, int> > tooNearCandidates;

	for (size_t i = 0; i < possibleMarkers.size(); i++) {

		const std::vector<cv::Point2f>& m1 = possibleMarkers[i];

		//calculate the average distance of each corner to the nearest corner of the other marker candidate
		for (size_t j = i + 1; j < possibleMarkers.size(); j++)
		{
			const std::vector<cv::Point2f>& m2 = possibleMarkers[j];
			float distSquared = 0;

			for (int c = 0; c < 4; c++)
			{
				cv::Point v = m1[c] - m2[c];
				distSquared += v.dot(v);
			}
			distSquared /= 4;

			if (distSquared < 100)
			{
				tooNearCandidates.push_back(std::pair<int, int>(i, j));
			}
		}
	}

	// Mark for removal the element of the pair with smaller perimeter
	std::vector<bool> removalMask(possibleMarkers.size(), false);

	for (size_t i = 0; i < tooNearCandidates.size(); i++)
	{
		float p1 = perimeter(possibleMarkers[tooNearCandidates[i].first]);
		float p2 = perimeter(possibleMarkers[tooNearCandidates[i].second]);

		size_t removalIndex;

		if (p1 > p2)
			removalIndex = tooNearCandidates[i].second;
		else
			removalIndex = tooNearCandidates[i].first;

		removalMask[removalIndex] = true;
	}

	// Return candidates
	detectedMarkers.clear();
	for (size_t i = 0; i < possibleMarkers.size(); i++)
	{
		if (!removalMask[i])
			detectedMarkers.push_back(possibleMarkers[i]);
	}

}

int markerDetector::hammDistMarker(cv::Mat bits)
{
	int dist = 0;

	for (int y = 0; y < 5; y++)
	{
		int minSum = static_cast<int>(1e5); //hamming distance to each possible word

		for (int p = 0; p < 5; p++)
		{
			int sum = 0;

			//now, count
			for (int x = 0; x < 5; x++)
				sum += bits.at<uchar>(y, x) == templateMarker[p * 5 + x] ? 0 : 1;

			if (minSum > sum)
				minSum = sum;
		}

		//do the and
		dist += minSum;
	}

	return dist;
}

void markerDetector::estimatePosition(std::vector<std::vector<cv::Point2f>>& detectedMarkers)
{
	for (size_t i = 0; i < detectedMarkers.size(); i++)
	{
		std::vector<cv::Point2f>& m = detectedMarkers[i];

		cv::Mat Rvec;
		cv::Mat_<float> Tvec;
		cv::Mat raux, taux;
		cv::solvePnP(m_markerCorners3d, m, camMatrix, distCoeff, raux, taux);
		raux.convertTo(Rvec, CV_32F);
		taux.convertTo(Tvec, CV_32F);

		cv::Mat_<float> rotMat(3, 3);
		cv::Rodrigues(Rvec, rotMat);

		// Copy to transformation matrix
		for (int col = 0; col < 3; col++)
		{
			for (int row = 0; row < 3; row++)
			{
				transformation.r().mat[row][col] = rotMat(row, col); // Copy rotation component
			}
			transformation.t().data[col] = Tvec(col); // Copy translation component
		}

		// Since solvePnP finds camera location, w.r.t to marker pose, to get marker pose w.r.t to the camera we invert it.
		transformation = transformation.getInverted();
	}
}

cv::Mat rotate(cv::Mat in)
{
	cv::Mat out;
	in.copyTo(out);
	for (int i = 0; i < in.rows; i++)
	{
		for (int j = 0; j < in.cols; j++)
		{
			out.at<uchar>(i, j) = in.at<uchar>(in.cols - j - 1, i);
		}
	}

	return out;
}

std::vector<std::vector<cv::Point2f>> markerDetector::recognizeMarkers(const cv::Mat& gray, std::vector<std::vector<cv::Point2f>>& detectedMarkers, bool show)
{
	std::vector<std::vector<cv::Point2f>> goodMarkers;

	for (size_t i = 0; i < detectedMarkers.size(); i++) {

		cv::Mat canonicalMarker;
		std::vector<cv::Point2f>& marker = detectedMarkers[i];

		// Find the perspective transfomation that brings current marker to rectangular form
		cv::Mat M = cv::getPerspectiveTransform(marker, m_markerCorners2d);

		// Transform image to get a canonical marker image
		cv::warpPerspective(gray, canonicalMarker, M, markerSize);

		//threshold image
		cv::threshold(canonicalMarker, canonicalMarker, 125, 255, cv::THRESH_BINARY | cv::THRESH_OTSU);

		//Read the code
		cv::Mat bitMatrix = cv::Mat::zeros(5, 5, CV_8UC1);

		//get information (for each innner square, determine if it is black or white)
		for (int y = 0; y < 5; y++)
			for (int x = 0; x < 5; x++)
			{
				int cellX = (x + 1) * cellSize;
				int cellY = (y + 1) * cellSize;
				cv::Mat cell = canonicalMarker(cv::Rect(cellX, cellY, cellSize, cellSize));

				int nZ = cv::countNonZero(cell);
				if (nZ > (cellSize * cellSize / 2))
					bitMatrix.at<uchar>(y, x) = 1;
			}

		//check all possible rotations
		cv::Mat rotations[4];
		int distances[4];
		rotations[0] = bitMatrix;
		distances[0] = hammDistMarker(rotations[0]);
		std::pair<int, int> minDist(distances[0], 0);

		for (int i = 1; i < 4; i++)
		{
			//get the hamming distance to the nearest possible word
			rotations[i] = rotate(rotations[i - 1]);
			distances[i] = hammDistMarker(rotations[i]);
			if (distances[i] < minDist.first)
			{
				minDist.first = distances[i];
				minDist.second = i;
			}
		}

		if (minDist.first == 0)
		{
			int nRotations = minDist.second;
			//sort the points so that they are always in the same order
			//no matter the camera orientation
			std::rotate(marker.begin(), marker.begin() + 4 - nRotations, marker.end());
			goodMarkers.push_back(marker);
		}
	}

	//Marker locaiton refinement
	if (goodMarkers.size() > 0)
	{
		std::vector<cv::Point2f> preciseCorners(4 * goodMarkers.size());
		for (size_t i = 0; i < goodMarkers.size(); i++)
		{
			std::vector<cv::Point2f>& marker = goodMarkers[i];
			for (int c = 0; c < 4; c++)
			{
				preciseCorners[i * 4 + c] = marker[c];
			}
		}

		cv::cornerSubPix(gray, preciseCorners, cvSize(5, 5), cvSize(-1, -
			1), cvTermCriteria(CV_TERMCRIT_ITER, 30, 0.1));

		//copy back
		for (size_t i = 0; i < goodMarkers.size(); i++)
		{
			std::vector<cv::Point2f>& marker = goodMarkers[i];
			for (int c = 0; c < 4; c++)
			{
				marker[c] = preciseCorners[i * 4 + c];
			}
		}
	}

	if (show)
	{
		cv::Mat imgShow;
		cv::cvtColor(gray, imgShow, cv::COLOR_GRAY2BGR);
		float thickness = 2;
		cv::Scalar color = cv::Scalar(0, 0, 255);
		for (size_t i = 0; i < goodMarkers.size(); i++) {
			std::vector<cv::Point2f> current_mark = goodMarkers[i];
			std::vector<cv::Point2i> int_current_mark;
			cv::Mat(current_mark).copyTo(int_current_mark);
			cv::line(imgShow, int_current_mark[0], int_current_mark[1], color, thickness, CV_AA);
			cv::line(imgShow, int_current_mark[1], int_current_mark[2], color, thickness, CV_AA);
			cv::line(imgShow, int_current_mark[2], int_current_mark[3], color, thickness, CV_AA);
			cv::line(imgShow, int_current_mark[3], int_current_mark[0], color, thickness, CV_AA);
		}
		imshow("mark", imgShow);
		imgShow = cv::Mat::zeros(cv::Size(640, 480), CV_8UC3);
	}

	return goodMarkers;
}

File: SmartSightBase/MarkerDetection/MarkerDetector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using OpenCvSharp;
using SmartSightBase.GeometryTypes;

namespace SmartSightBase
{
    public class MarkerDetector
    {
        private float mMinContourLengthAllowed = 100.0f;
        private Size mMarkerSize = new Size(100,100);
        private int mCellSize = 100 / 7;
        private Mat camMatrix = new Mat();
        private Mat mDistCoeff = new Mat();
        private Mat mGrayscale = new Mat();
        private Mat mThresholdImg = new Mat();

        private List<List<Point2f>> mDetectedMarkers = new List<List<Point2f>>();
        private List<List<Point>> mContours = new List<List<Point>>();
        private List<Point3f> mMarkerCorners3d = new List<Point3f>();
        private List<Point2f> mMarkerCorners2d = new List<Point2f>();

        public event EventHandler MarkerDetected = (s, e) => { };

        public List<List<Point2f>> GoodMarkers { get; set; } = new List<List<Point2f>>();

        public Transformation Transformation { get; set; }

        public int[] TemplateMarker { get; set; }

        public Mat DetectedMarkerImg { get; set; } = new Mat();

        public MarkerDetector(CameraCalibration calibration, Monitor monitor)
        {
            camMatrix = new Mat(3, 3, MatType.CV_32F, calibration.GetIntrinsic().Data[0]);
            mDistCoeff = new Mat(4, 1, MatType.CV_32F, calibration.GetDistorsion().Data[0]);

            mMarkerCorners3d.Add(new Point3f(-0.5f, -0.5f, 0));
            mMarkerCorners3d.Add(new Point3f(+0.5f, -0.5f, 0));
            mMarkerCorners3d.Add(new Point3f(+0.5f, +0.5f, 0));
            mMarkerCorners3d.Add(new Point3f(-0.5f, +0.5f, 0));

            mMarkerCorners2d.Add(new Point2f(0, 0));
            mMarkerCorners2d.Add(new Point2f(((float)mMarkerSize.Width - 1), 0));
            mMarkerCorners2d.Add(new Point2f(((float)mMarkerSize.Width - 1), (float)(mMarkerSize.Height - 1)));
            mMarkerCorners2d.Add(new Point2f(0, (float)(mMarkerSize.Height - 1)));

            this.TemplateMarker = new int[25]{ 1, 1, 1, 0, 1,
                                               0, 0, 0, 1, 0,
                                               1, 1, 1, 1, 0,
                                               1, 1, 1, 1, 0,
                                               1, 1, 1, 0, 0 };
        }

        public void FindMarkers(Monitor monitor, bool showImg)
        {
            this.PrepareImage(monitor.CameraImg, mGrayscale);

            this.PerformThreshold(mGrayscale, mThresholdImg);

            this.FindContours(mThresholdImg, mContours, mGrayscale.Cols / 5);

            this.FindCandidates(mContours, mDetectedMarkers);

            this.GoodMarkers = this.RecognizedMarkers(mGrayscale, mDetectedMarkers, showImg);

            this.EstimatePosition(this.GoodMarkers);
        }

        protected void PrepareImage(Mat bgrMat, Mat grayscale)
        {
            Cv2.CvtColor(bgrMat, grayscale, ColorConversionCodes.BGR2GRAY);
        }

        protected void PerformThreshold(Mat grayscale, Mat thresholdImg)
        {
            Cv2.AdaptiveThreshold(grayscale, thresholdImg, 255, AdaptiveThresholdTypes.GaussianC, ThresholdTypes.BinaryInv, 7, 7);
        }

        protected void FindContours(Mat thresholdImg, List<List<Point>> contours, int minContourPointsAllowed)
        {
            Point[][] allContours;
            HierarchyIndex[] hierarchyIndexes;
            Cv2.FindContours(thresholdImg, out allContours, out hierarchyIndexes, RetrievalModes.List, ContourApproximationModes.ApproxNone);

            contours.Clear();

            for (var i = 0; i < allContours.Length; i++)
            {
                var contourSize = allContours[i].Length;
                if (contourSize > minContourPointsAllowed)
                {
                    contours.Add(allContours[i].ToList());
                }
            }
        }

        protected float Perimeter(List<Point2f> list)
        {
            float sum = 0, dx, dy;

            for (var i = 0; i < list.Count; i++)
            {
                var i2 = (i + 1) % list.Count;

                dx = list[i].X - list[i2].X;
                dy = list[i].Y - list[i2].Y;

                sum += (float)Math.Sqrt(dx * dx + dy * dy);
            }

            return sum;
        }

        protected void FindCandidates(List<List<Point>> contours, List<List<Point2f>> marks)
        {
            //var approxCurve = new List<Point>();
            var possibleMarkers = new List<List<Point2f>>();

            // For each contour, analyze if it is a parallelepiped likely to be the	marker
            for (var i = 0; i < contours.Count; i++)
            {
                //Approximate to a polygon
                var eps = contours[i].Count * 0.05;

                try
                {
                    var inputArray = InputArray.Create(contours[i]);
                    var outputArray = OutputArray.Create(contours[i]);
                }
                catch(ArgumentNullException)
                {
                    // Chomp chomp, no idea
                }

                var approxCurve = Cv2.ApproxPolyDP(contours[i], eps, true).ToList();

                //Cv2.ApproxPolyDP(inputArray, outputArray, eps, true);

                // We interested only in polygons that contains only four points
                if (approxCurve.Count != 4)
                    continue;

                // And they have to be convex
                if (!Cv2.IsContourConvex(approxCurve))
                    continue;

                // Ensure that the distance between consecutive points is large enough
                var minDist = float.MaxValue;

                for (var j = 0; j < 4; j++)
                {
                    var side = approxCurve[j] - approxCurve[(j + 1) % 4];
                    var squaredSideLength = (float)side.DotProduct(side);
                    minDist = Math.Min(minDist, squaredSideLength);
                }

                // Check that distance is not very small
                if (minDist < mMinContourLengthAllowed)
                    continue;

                // All tests are passed. Save marker candidate:
                var m = new List<Point2f>();
                for (var j = 0; j < 4; j++)
                    m.Add(new Point2f(approxCurve[j].X, approxCurve[j].Y));

                // Sort the points in anti-clockwise order
                // Trace a line between the first and second point.
                // If the third point is at the right side, then the points are anticlockwise
                var v1 = m[1] - m[0];
                var v2 = m[2] - m[0];

                double o = (v1.X * v2.Y) - (v1.Y * v2.X);
                if (o < 0.0) //if the third point is in the left side,	then sort in anti - clockwise order
                {
                    var temp = m[1];
                    var temp2 = m[3];
                    m[1] = temp2;
                    m[3] = temp;
                }

                possibleMarkers.Add(m);
            }

            // Remove these elements which corners are too close to each other.
            // First detect candidates for removal:
            var tooNearCandidates = new List<Tuple<int, int>>();
            for (var i = 0; i < possibleMarkers.Count; i++)
            {
                var m1 = possibleMarkers[i];

                //calculate the average distance of each corner to the nearest corner of the other marker candidate
                for (var j = i + 1; j < possibleMarkers.Count; j++)
                {
                    var m2 = possibleMarkers[j];
                    float distSquared = 0;

                    for (var c = 0; c < 4; c++)
                    {
                        var v = m1[c] - m2[c];
                        distSquared += (float)v.DotProduct(v);
                    }

                    distSquared /= 4;

                    if (distSquared < 100)
                    {
                        tooNearCandidates.Add(new Tuple<int, int>(i, j));
                    }
                }
            }

            // Mark for removal the element of the pair with smaller perimeter
            var removalMask = Enumerable.Repeat(false, possibleMarkers.Count).ToList();

            for (var i = 0; i < tooNearCandidates.Count; i++)
            {
                var p1 = this.Perimeter(possibleMarkers[tooNearCandidates[i].Item1]);
                var p2 = this.Perimeter(possibleMarkers[tooNearCandidates[i].Item2]);
                int removalIndex;

                if (p1 > p2)
                    removalIndex = tooNearCandidates[i].Item2;
                else
                    removalIndex = tooNearCandidates[i].Item1;

                removalMask[removalIndex] = true;
            }

            // Return candidates
            mDetectedMarkers.Clear();
            for (var i = 0; i < possibleMarkers.Count; i++)
            {
                if (!removalMask.ElementAt(i))
                    mDetectedMarkers.Add(possibleMarkers[i]);
            }
        }

        protected List<List<Point2f>> RecognizedMarkers(Mat gray, List<List<Point2f>> detectedMarkers, bool showImg)
        {
            this.GoodMarkers = new List<List<Point2f>>();
            for (var i = 0; i < detectedMarkers.Count; i++)
            {
                var canonicalMarker = new Mat();
                var marker = detectedMarkers[i];

                // Find the perspective transfomation that brings current marker to rectangular form
                var m = Cv2.GetPerspectiveTransform(marker, mMarkerCorners2d);

                // Transform image to get a canonical marker image
                Cv2.WarpPerspective(gray, canonicalMarker, m, mMarkerSize);

                //threshold image
                Cv2.Threshold(canonicalMarker, canonicalMarker, 125, 255, ThresholdTypes.Binary | ThresholdTypes.Otsu);

                //Read the code
                Mat bitMatrix = Mat.Zeros(5, 5, MatType.CV_8UC1);

                //get information (for each innner square, determine if it is black or white)
                for (var y = 0; y < 5; y++)
                    for (var x = 0; x < 5; x++)
                    {
                        var cellX = (x + 1) * mCellSize;
                        var cellY = (y + 1) * mCellSize; 
                        var cell = canonicalMarker.SubMat(new Rect(cellX, cellY, mCellSize, mCellSize));

                        var nZ = Cv2.CountNonZero(cell);
                        if (nZ > (mCellSize * mCellSize / 2))
                            bitMatrix.Set<byte>(y, x, 1);
                    }

                //check all possible rotations
                var rotations = new Mat[4];
                var distances = new int[4];
                rotations[0] = bitMatrix;
                distances[0] = this.HammDistMarker(rotations[0]);
                var minDist = new Tuple<int, int>(distances[0], 0);

                for (var j = 1; j < 4; j++)
                {
                    //get the hamming distance to the nearest possible word
                    rotations[j] = this.Rotate(rotations[j - 1]);
                    distances[j] = this.HammDistMarker(rotations[j]);

                    if (distances[j] < minDist.Item1)
                    {
                        minDist = new Tuple<int, int>(distances[j], j);
                    }
                }

                if (minDist.Item1 == 0)
                {
                    var nRotations = minDist.Item2;
                    //sort the points so that they are always in the same order
                    //no matter the camera orientation
                    var rotated = marker.Skip(marker.IndexOf(marker.First()) + 4 - nRotations).Concat(marker.Take(marker.IndexOf(marker.First()) + 4 - nRotations)).ToList();

                    this.GoodMarkers.Add(rotated);
                }

            }

            //Marker locaiton refinement
            if (this.GoodMarkers.Count > 0)
            {
                var preciseCorners = Enumerable.Repeat(new Point2f(), 4 * this.GoodMarkers.Count).ToList();

                for (var i = 0; i < this.GoodMarkers.Count; i++)
                {
                    var marker = this.GoodMarkers[i];

                    for (var c = 0; c < 4; c++)
                    {
                        preciseCorners[i * 4 + c] = marker[c];
                    }

                    Cv2.CornerSubPix(gray, preciseCorners, new Size(5, 5), new Size(-1, -
                            1), new TermCriteria(CriteriaType.MaxIter, 30, 0.1));

                    //copy back
                    for (var j = 0; j < this.GoodMarkers.Count; j++)
                    {
                        var marker2 = this.GoodMarkers[i];
                        for (var c = 0; c < 4; c++)
                        {
                            marker2[c] = preciseCorners[i * 4 + c];
                        }
                    }
                }

                if (showImg)
                {
                    Cv2.CvtColor(gray, this.DetectedMarkerImg, ColorConversionCodes.GRAY2BGR);
                    var thickness = 2;
                    var color = new Scalar(0, 0, 255);

                    for (var i = 0; i < this.GoodMarkers.Count; i++)
                    {
                        var current_mark = this.GoodMarkers[i];
                        var int_current_mark = Enumerable.Repeat(new Point2f(), (this.GoodMarkers[i].Count)).ToArray();
                        current_mark.CopyTo(int_current_mark);
                        Cv2.Line(this.DetectedMarkerImg, (int)int_current_mark[0].X, (int)int_current_mark[0].Y, (int)int_current_mark[1].X, (int)int_current_mark[1].Y, color, thickness, LineTypes.AntiAlias);
                        Cv2.Line(this.DetectedMarkerImg, (int)int_current_mark[1].X, (int)int_current_mark[1].Y, (int)int_current_mark[2].X, (int)int_current_mark[2].Y, color, thickness, LineTypes.AntiAlias);
                        Cv2.Line(this.DetectedMarkerImg, (int)int_current_mark[2].X, (int)int_current_mark[2].Y, (int)int_current_mark[3].X, (int)int_current_mark[3].Y, color, thickness, LineTypes.AntiAlias);
                        Cv2.Line(this.DetectedMarkerImg, (int)int_current_mark[3].X, (int)int_current_mark[3].Y, (int)int_current_mark[0].X, (int)int_current_mark[0].Y, color, thickness, LineTypes.AntiAlias);
                    }
                }

                MarkerDetected.Invoke(this, new EventArgs());
            }

            return this.GoodMarkers;
        }

        protected Mat Rotate(Mat matrix)
        {
            var outMatrix = new Mat();
            matrix.CopyTo(outMatrix);

            for (var i = 0; i < matrix.Rows; i++)
            {
                for (var j = 0; j < matrix.Cols; j++)
                {
                    outMatrix.Set<byte>(i, j, matrix.At<byte>(matrix.Cols - j - 1, i));
                }
            }

            return outMatrix;
        }


        protected int HammDistMarker(Mat bits)
        {
            var dist = 0;

            for (var y = 0; y < 5; y++)
            {
                var minSum = (int)1e5; //hamming distance to each possible word

                for (var p = 0; p < 5; p++)
                {
                    var sum = 0;

                    //now, count
                    for (var x = 0; x < 5; x++)
                    {
                        sum += bits.At<byte>(y, x) == this.TemplateMarker[p * 5 + x] ? 0 : 1;
                    }

                    if (minSum > sum)
                        minSum = sum;
                }

                //do the and
                dist += minSum;
            }

            return dist;
        }

        protected void EstimatePosition(List<List<Point2f>> detectedMarkers)
        {
            for (var i = 0; i < detectedMarkers.Count; i++)
            {
                var m = detectedMarkers[i];
                Mat Rvec = new Mat();
                Mat Tvec = new Mat();
                Mat raux = new Mat(), taux = new Mat();
                var inputArrayCorners = InputArray.Create(mMarkerCorners3d);
                var inputArrayImage = InputArray.Create(m);
                var outputArrayRaux = OutputArray.Create(raux);
                var outputArrayTaux = OutputArray.Create(taux);

                Cv2.SolvePnP(inputArrayCorners, inputArrayImage, camMatrix, mDistCoeff, raux, taux);

                raux.ConvertTo(Rvec, MatType.CV_32F);
                taux.ConvertTo(Tvec, MatType.CV_32F);

                var rotMat = new Mat(3, 3, MatType.CV_32F);
                Cv2.Rodrigues(Rvec, rotMat);

                // Copy to transformation matrix
                for (var col = 0; col < 3; col++)
                {
                    for (var row = 0; row < 3; row++)
                    {
                        if (this.Transformation == null)
                            this.Transformation = new Transformation();

                        this.Transformation.Rotation().Mat[row, col] = rotMat.Get<float>(row, col); // Copy rotation component
                    }

                    this.Transformation.Translation().Data[col] = Tvec.Get<float>(col); // Copy translation component
                }

                // Since solvePnP finds camera location, w.r.t to marker pose, to get marker pose w.r.t to the camera we invert it.
                this.Transformation = this.Transformation.GetInverted();
            }
        }
    }
}
